src/Platform/OpenGL/Shaders/god_ray_fragment.glsl

```diff
@@ -0,0 +1,95 @@
+#include "Library/directional_light.glsl"
+
+in vec2 TexCoord;
+out vec4 OutColor;
+
+uniform sampler2D albedoTex;
+uniform sampler2D normalTex;
+uniform sampler2D materialTex;
+uniform sampler2D depthTex;
+
+struct Camera
+{
+    vec3 position;
+    mat4 invViewProjMatrix;
+    mat4 viewProjMatrix;
+};
+
+uniform sampler2D cameraOutput;
+uniform Camera camera;
+
+uniform int lightCount;
+uniform DirLight lights[MaxDirLightCount];
+uniform sampler2D lightDepthMaps[MaxDirLightCount];
+
+uniform float maxSteps;
+uniform float sampleStep;
+uniform float stepIncrement;
+
+//sample only one time without extra interpolation
+float godRayShadowFactor2D(vec3 coords, sampler2D depthMap, vec4 textureLimitsXY, float bias)
+{
+    float compare = coords.z - bias;
+    
+    const float lod = getTextureLodLevel(coords.xy);
+    vec2 texelSize = textureSize(depthMap, 0);
+    vec2 texelSizeInv = 1.0 / texelSize;
+    vec2 pixelPos = coords.xy * texelSize + vec2(0.5);
+    vec2 fracPart = fract(pixelPos);
+    vec2 startTexel = (pixelPos - fracPart) * texelSizeInv;
+ 	float res = sampleShadowMap(depthMap, startTexel + vec2( 0.0,  0.0) * texelSizeInv, lod, compare); 
+	return pow(clamp(res, 0.0, 1.0), 8.0);
+}
+float godRayShadowSampler(vec3 position, DirLight light, sampler2D shadowMap)
```

Comment:
can you explain a bit how it differs from default engine cascade sampler?

Reply:
Texture sampling is much more expensive than arithmetic operation. The default one gives better results however it requires many samples. For this feature I need to know is if a given place is occluded. I don't care about the influence of its neighbors. So I built its own sampler to achieve better performance.